core/lib/zksync_core/src/api_server/web3/namespaces/zks.rs

```diff
@@ -499,20 +499,33 @@ impl ZksNamespace {
         address: Address,
         keys: Vec<H256>,
         l1_batch_number: L1BatchNumber,
-    ) -> Result<Proof, Web3Error> {
+    ) -> Result<Option<Proof>, Web3Error> {
         self.state.start_info.ensure_not_pruned(l1_batch_number)?;
         let hashed_keys = keys
             .iter()
             .map(|key| StorageKey::new(AccountTreeId::new(address), *key).hashed_key_u256())
             .collect();
-        let proofs = self
+        let tree_api = self
             .state
             .tree_api
-            .as_ref()
-            .ok_or(Web3Error::TreeApiUnavailable)?
-            .get_proofs(l1_batch_number, hashed_keys)
-            .await
-            .context("get_proofs")?;
+            .as_deref()
+            .ok_or(Web3Error::TreeApiUnavailable)?;
+        let proofs_result = tree_api.get_proofs(l1_batch_number, hashed_keys).await;
+        let proofs = match proofs_result {
+            Ok(proofs) => proofs,
+            Err(TreeApiError::NotReady) => return Err(Web3Error::TreeApiUnavailable),
+            Err(TreeApiError::NoVersion(err)) => {
+                return if err.missing_version > err.version_count {
+                    // FIXME: should we distinguish the case when the tree is catching up?
```

Comment:
We could fetch the latest L1 batch number from Postgres and compare it to `err.version_count`. If it's greater, then the Merkle tree is catching up. Not sure if it makes sense from the DevX perspective, or would just create an additional error that clients would need to handle.

Reply:
I think we can leave it as is and maybe revisit it later if someone complains.